### PR TITLE
feat: support codeowners in catalog-import

### PR DESCRIPTION
**Pull request by andrewthauer**
## Hey, I just made a Pull Request!

The new `catalog-import` imposed a requirement to supply an `owner` property for each entity. This however, breaks the usage of the existing `CodeOwnersProcessor` when using the import feature.

This adds a new `Use CODEOWNERS File` checkbox option that allows a user to explicitly indicate they want to use the `CodeOwnersProcessor` and omit the `owner` field. Selected the checkbox will hide the `owner` group selector and therefore ignore the validation for that form field. Otherwise validation will require the `owner` to be specified.

**Next Steps** - Make this checkbox automatically selected if there is a CODEOWNERS file found at the target location. Or possibly a combination of a config flag saying `preferCodeowners` + the presence of the file at the target.

##### Without `Use CODEOWNERS File` selected:

image

image

##### With `Use CODEOWNERS File` selected:

image

image

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4588-godojucjvbjgdaqd24lf5njfcbjgzz1k.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602f27a0f285afad3392bf52

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4588-godojucjvbjgdaqd24lf5njfcbjgzz1k.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602f27a0f285afad3392bf52

**Comment by Fox32** [comment]
I wonder if this should be either a configuration option or use the ability to replace the dialog with a custom one. I thing @dhenneke took care during the refactor that you can replace a lot of things.

How about checking whether a codeowner file is present and then showing the option instead?

**Comment by freben** [comment]
> How about checking whether a codeowner file is present and then showing the option instead?

That is the design I've been proposing. The backend should be able to detect the presence of a codeowners file and tell the user literally that "hey I found that codeowners would set this to X so I _chose to leave this empty_ and suggest you keep it empty, but you can fill in an override if you want to"

**Comment by dhenneke** [comment]
Yes I didn't take this codeowners thing into account because I was not really aware of it / hoped it will not be a problem :/

I would prefer to not add too much workarounds into the component but to move this out of it. I suspect that a backstage instance either always uses codeowners or uses none? However this might open the possibility that the user creates a broken PR because the codeowners file is missing for some reason and we don't recognized this in the form.

There should be multiple alternatives here:

1. Override the generate function and provide a form without the owner field or a non-required one (https://github.com/backstage/backstage/blob/cf9a229a3e72db24e848f3e000c8a2850224e344/plugins/catalog-import/src/components/ImportStepper/defaults.tsx#L236). But I guess the overriding part is not yet exposed from the ImportComponentPage.
2. If we think that this is a very common case, it could also be a setting in https://github.com/backstage/backstage/blob/cf9a229a3e72db24e848f3e000c8a2850224e344/plugins/catalog-import/src/components/ImportStepper/defaults.tsx#L31 that removes the required configuration in the form.

In both cases, this code should be changed to handle an undefined or an empty string owner: https://github.com/backstage/backstage/blob/cf9a229a3e72db24e848f3e000c8a2850224e344/plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx#L47. Not sure what useForm returns for empty values.

What came to my mind: You could also add the option to pass a default owner to the form and then override the api that includes the logic that you wrote here and forwards it to the original create-or code.


Either option shouldn't substantially differ in effort.

**Comment by dhenneke** [comment]
> > How about checking whether a codeowner file is present and then showing the option instead?
> 
> 
> 
> That is the design I've been proposing. The backend should be able to detect the presence of a codeowners file and tell the user literally that "hey I found that codeowners would set this to X so I _chose to leave this empty_ and suggest you keep it empty, but you can fill in an override if you want to"

That's also a good alternative. A quick fix could be to change the frontend form to support such and either directly implement the backend part or let the integrator override the CatalogInportAPI for now to hardcore the setting

**Comment by Fox32** [comment]
As codeowner seems to be active by default, I think first class support without configuration would be neat. We could:

* Call the location analyzer in the backend from the analyze step 
* Remove the actual logic from `CodeOwnerProcessor` and move it into the integration package.
* Location analyzer calls this code owners evaluation code
* And optional returns a possible owner if one was found and that it comes from codeowners?
* The UI uses the owner in the `no-location` flow and doesn't provide the option to override it but just display that it's deduced fro m codeowners (maybe even with the current owner)?

**Comment by freben** [comment]
The integration package is frontend-and-backend compatible, i won't be able to house logic like that.

Well I think the end result importantly must be that the owner IS not set at all in the yaml file (because it gets populated dynamically on each refresh). It should not be filled in with a snapshot of what the codeowners currently says; that kinda defeats the flexible dynamic codeowners functionality.

**Comment by freben** [comment]
Essentially I think that the response from the backend analysis should be:

- A partial Entity with a lot of stuff filled in but maybe not all
- A list of fields (probably as a jsonschema form) that must be presented to the user if it isn't empty.

The list of fields will have descriptive texts per entry, that is shown under the form field. Some will be optional, some will not. The frontend code needs to be smart enough to take the inputs in those fields, and put them at the right places in the entity yaml blob.

**Comment by andrewthauer** [comment]
The recent changes are fanstastic. Thank you @dhenneke for contributing them :smile:

However, unfortunately they did not completely factor in the codeowner use case or usage of a custom location analyzer which we use both of to customize the entity files generated. Also see #4590. This means we now need to pick a random owner + go into the PR and edit the submitted PR to cleanup the changes.

Currently our custom LocationAnalyzer does the following differently then the default one:

- omits the owner (as done by the default), but this needs to be preserved for the processor to work
- fetches the description from the repo and populates that
- assumes `lifecycle` = `production` as it almost always is
- does some smart naming analysis and sets the `spec.type`. (e.g. `something-service` = `service` or `something-ui` = `website` + a few other conventional things)

Most of these features from the custom location analyzer still work except for the `owner` field being required and what #4590 is meant to address to help maintain consistency across our catalog files by not introducing a new unused key for `namespace`

I agree this probably could and should be automated. This was meant to be a quick fix to make the experience in the short term a little easier for us as we are making a push to get teams to fill the catalog. Unfortunately I did not notice this earlier  and also missed the release window :sad:. Hoping, everyone might be ok to accept this so we can use a nightly build in the short term while I address the concerns & ideas in follow up PR?

**Comment by andrewthauer** [comment]
I also agree we should have first class support for CODEOWNER files. This functionality is key for us and we've been bitten a couple times as it's not a well known use case and therefore not always factored into changes. The fact that functionality requires fields to be omitted for later processing is not obvious to both end users (catalog maintainers) & code maintainers.

At one point I suggested using a special key like `owner: { $codeowner }`. I'm wondering if something like that might be a better approach in the long term. However, the same pattern of omitting fields could be useful for other things so maybe a general keyword could be more applicable. I think this is mostly a problem for things that are required fields.

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx
Comment (dhenneke) [inline plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx:121]: The problem I see now is that this option is available by default even though they don't use codeowners at all and/or don't use the GitHub groups in backstage. So this might need additional documentation or it could lead to confused users who create catalog files which can't be imported due to missing required fields.

Comment (andrewthauer) [reply plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx:121]: That's a fair point. A config flag would be a quick fix as well but introduce something that would need to be deprecated for a more automated solution. Let me think on this a little.

**Comment by dhenneke** [comment]
> Hoping, everyone might be ok to accept this as is (or minor tweaks) so we can use a nightly build in the short term while I address the concerns & ideas in follow up PR?

You could also put the code you wrote into a api wrapper that changes the results/inputs of the analyze and create-pr functions. Should lead to the same behavior but doesn't require upstream changes. Could buy us some time to fix this right without adding already deprecated behavior (see https://github.com/backstage/backstage/blob/4a15d5da5e695c63a7a148aaac4714d4f8dd9622/plugins/catalog-import/src/api/CatalogImportApi.ts#L42)

**Comment by andrewthauer** [comment]
> > Hoping, everyone might be ok to accept this as is (or minor tweaks) so we can use a nightly build in the short term while I address the concerns & ideas in follow up PR?
> 
> You could also put the code you wrote into a api wrapper that changes the results/inputs of the analyze and create-pr functions. Should lead to the same behavior but doesn't require upstream changes. Could buy us some time to fix this right without adding already deprecated behavior (see
> 
> https://github.com/backstage/backstage/blob/4a15d5da5e695c63a7a148aaac4714d4f8dd9622/plugins/catalog-import/src/api/CatalogImportApi.ts#L42
> )

Oh nice. I didn't realize that was there. Ok, that seems like a better option for time being. Maybe I should close this and open an issue / RFC around this in general. There's a few things at play here we've discussed, so I'll need to think about untangling the concerns a bit.

**Comment by freben** [comment]
I think the analysis step on the backend requires some love and attention overall. Maybe we can find a way to make this offering a bit more complete overall - along the lines of what I mentioned elsewhere, i.e. that the analysis step is smart enough to conclude all of this and conveying it to the frontend, and the frontend is smart enough to ask the user the right questions given that.

**Comment by freben** [comment]
Let me know how you want to proceed.

**Comment by andrewthauer** [comment]
Well, changing out the `CataligImportApi` requires copying over a boat load of code from a few different files. It's managable in the short term, but not ideal.

**Comment by dhenneke** [comment]
Hmm. I assumed you could extend the existing implementation and just add/remove the data while doing super.* calls to reuse the existing logic. But I hadn’t tried that yet.

**Comment by andrewthauer** [comment]
> Hmm. I assumed you could extend the existing implementation and just add/remove the data while doing super.* calls to reuse the existing logic. But I hadn’t tried that yet.

Yeah, was just trying to get this to work quickly work, but I'll give that a shot.

**Comment by andrewthauer** [comment]
Unfortunately, overriding the `CataligImportApi` approach does very little to help since the `submitPullRequest` interface method takes the already generated file contents and would require parsing the yaml and then re-building it. This is past what I'm willing to take on as custom code just to work around this unfortunately. I think I'm back to hoping we can include this as a short term fix.

The biggest concern would be that having this option might mislead people to use it without having an actual codeowner file. This would then cause the validation to fail on the backend and not be noticed. I could look at re-working it to have a checkbox with some disclaimer text. Something like:

[  ] Use CODEOWNERS file from repo as owner (_this will attempt to fill in the spec.owner field automatically from a `CODEOWNERS` file within the repo. If one is not found this could cause the import to file)_
      
Checking this option would disable the group selector and have a similar behaviour that I'm proposing. Later on we can extend this to be automatically visible and checked when there is a code owner file. Or possibly only visible when a config flag is enabled and there is a file. Thoughts?

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4588-godojucjvbjgdaqd24lf5njfcbjgzz1k.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602f27a0f285afad3392bf52

**Comment by andrewthauer** [comment]
Ok, so I've added a new commit which changes the original group select option into a checkbox control. This seems to work pretty well and should be both more obvious behaviour and future proof for a more smart version of this of this based on analysis of the target. I also updated the description of the PR to reflect this + screen shots.

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/PreparePullRequestForm.tsx
Comment (dhenneke) [inline plugins/catalog-import/src/components/StepPrepareCreatePullRequest/PreparePullRequestForm.tsx:35]: The values prop already provides the value of watch(). Also why do we need formState?

File: plugins/catalog-import/src/components/ImportStepper/defaults.tsx
Comment (dhenneke) [inline plugins/catalog-import/src/components/ImportStepper/defaults.tsx:252]: This should be register() which should also setup the binding to the field that is specified by the name prop. You then shouldn't need to have the onchange handler.

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/CheckboxField.tsx
Comment (dhenneke) [inline plugins/catalog-import/src/components/StepPrepareCreatePullRequest/CheckboxField.tsx:35]: Not sure if we really need this extra component. There are some quite complex type definitions and I'm not sure if it's worth doing that. Why not add the structure to the defaults.tsx? The Autocomplete component is special because it is not compatible to the form library by default which I would expect the checkbox to be.

File: plugins/catalog-import/src/components/ImportStepper/defaults.tsx
Comment (dhenneke) [inline plugins/catalog-import/src/components/ImportStepper/defaults.tsx:257]: Ah I see what you are doing here. You are bypassing the react-hook-form logic. We should definitely fix this later so we don't get splitted logic with unclear side effects.

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx
Comment (dhenneke) [inline plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx:89]: Why not useCodeowners ? {} : { owner }?

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/PreparePullRequestForm.tsx
Comment (andrewthauer) [reply plugins/catalog-import/src/components/StepPrepareCreatePullRequest/PreparePullRequestForm.tsx:35]: Forgot to remove. My first time using `react-hook-form` so I was attempting to use it to toggle the button disabled property using `isValid` rather then the multiple checks it's using now. I'm used to Angular forms where there are more concepts like `pristine` and an aggregate `touched` which didn't work out. I essentially wanted the `submit` button to be enabled until the form was dirtied or invalid. Couldn't get the right mix to work. I'll remove this, unless you have an alternative suggestion.

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx
Comment (andrewthauer) [reply plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx:89]: I could, but was just avoiding bringing the that concept into this pure function.

File: plugins/catalog-import/src/components/ImportStepper/defaults.tsx
Comment (andrewthauer) [reply plugins/catalog-import/src/components/ImportStepper/defaults.tsx:257]: Yeah, I needed some way of resetting the owner. Open to other ideas for this PR as long as they don't require a large amount of refactoring :)

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/CheckboxField.tsx
Comment (andrewthauer) [reply plugins/catalog-import/src/components/StepPrepareCreatePullRequest/CheckboxField.tsx:35]: The main reason I did this was to to ecapsulate additional disclaimer text. I originally tried putting in an warning icon and also wrapping the text under the label, but the alignment starts looking funny. So I ended up falling back to just planning it under the checkbox and label altogether to avoid spending too much time styling. But yeah, now it seems a bit unnecessary for a different component.

File: plugins/catalog-import/src/components/ImportStepper/defaults.tsx
Comment (andrewthauer) [reply plugins/catalog-import/src/components/ImportStepper/defaults.tsx:252]: Sure, if I removed the wrapper component then that would work inline here. 👍

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4588-godojucjvbjgdaqd24lf5njfcbjgzz1k.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602f27a0f285afad3392bf52

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4588-godojucjvbjgdaqd24lf5njfcbjgzz1k.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602f27a0f285afad3392bf52

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/PreparePullRequestForm.tsx
Comment (dhenneke) [reply plugins/catalog-import/src/components/StepPrepareCreatePullRequest/PreparePullRequestForm.tsx]: That's a tricky one. The dirty check is implemented this way for the input-url-form but I couldn't get it to work in this create-or-form. I struggled with the fact that it is using default values for some fields and that non-touched defaults are interpreted as them being not dirty. So this would break if the defaults are already valid and I skipped this behavior.

**Review by dhenneke — APPROVED** [review]
Nice 👍

Could work for now but needs some rework when the codeowners functionality is more recognized as a core feature than it is now. What I don't like about this component (even without your change) is the need for a lot of documentation since the process itself creates questions and has many alternatives. I can imagine that this current form will rise a lot of questions from users to the integrators if they e.g. don't use codeowners and the users are not familiar with backstage internals. Or a lot of discussions at your company about why there is an author field if you shouldn't use it and why backstage not directly drop the custom author field since it's obvious that codeowners are more convenient.

That reminds me that I wanted to raise an issue about a reworked entrypoint into the whole create-new/add-existing experience based on ideas of @freben...

**Comment by andrewthauer** [comment]
@backstage/maintainers - Any chance I can get a maintainer review/approval on this 😄 ?

**Comment by andrewthauer** [comment]
@freben - Any chance you could have a once over again on this PR 😄 ?

**Review by freben — APPROVED** [review]
Yep, good enough for now to unblock the use case, then we have to tweak how the backend-frontend interactions look.

**Pull request merged**
